README.md
# koa-omnibus

[![Travis CI badge](https://travis-ci.org/hagemt/node-koa-omnibus.svg?branch=master)](https://travis-ci.org/hagemt/node-koa-omnibus)
[![Test Coverage](https://api.codeclimate.com/v1/badges/c5903ebb2706657b1820/test_coverage)](https://codeclimate.com/github/hagemt/node-koa-omnibus/test_coverage)
[![Maintainability](https://api.codeclimate.com/v1/badges/c5903ebb2706657b1820/maintainability)](https://codeclimate.com/github/hagemt/node-koa-omnibus/maintainability)
[![Greenkeeper badge](https://badges.greenkeeper.io/hagemt/node-koa-omnibus.svg)](https://account.greenkeeper.io/account/hagemt)

I wouldn't use this in production until it's more battle-tested. (at least v1)

It does exactly everything you want an API server to do, and nothing more.

Put this at the front of your `koa@^2` middleware chain and get hacking.

## What do you mean? I want to try before I buy.

Clone, then: `npm install && npm install koa && npm run demo` or:

Slap this in `node server.js | bunyan` after `npm install koa koa-omnibus`:

```
const omnibus = require('koa-omnibus')
omnibus.createApplication().listen()
// or: application.use(omnibus())
```

## Defaults not good enough? Okay...

All dependencies are included only for defaults; you can override anything you want.

```
omnibus({

	// simple constants: (all Numbers are in milliseconds)
	headers: { timing:String, tracking:String }, // defaults: 'X-Response-Time' and 'X-Request-ID'
	limits: Object { age:Number, max:Number, next:Number, rpm:Number }, // 60000/1000000/60000/1000
	namespace: String, // default: 'omnibus' (if set false-y, context.state decorated directly)

	// middleware factories (unary):
	limitRate: options:Object => middleware:AsyncFunction, // 429 Too Many Requests
	limitTime: options:Object => middleware:AsyncFunction, // 408 Request Timeout

	// redaction transforms (binary):
	redactedError: (options, context) => error, // do whatever transform(s) you want here
	redactedRequest: (options, context) => _.omit(context.request, ['header']), // " here
	redactedResponse: (options, context) => _.omit(context.response, ['header']), // " here

	// object factories (ternary):
	targetError: (options, context, error) => error, // not redacted (default: Boom)
	targetHeaders: (options, context, string) => Object { [string]: "$(uuidgen -t)" },
	targetLogger: (options, context, object) => log, // will include tracking headers
	targetObject: (options, context, object) => object, // context[namespace || 'state']

})
```

## Oh, yeah. One more thing:

This doesn't include any of these modules, but you can `.use` them, as you wish:

* https://www.npmjs.com/package/koa-bodyparser
* https://www.npmjs.com/package/koa-compress
* https://www.npmjs.com/package/koa-etag
* https://www.npmjs.com/package/koa-favicon
* https://www.npmjs.com/package/koa-helmet
* https://www.npmjs.com/package/koa-logger
* https://www.npmjs.com/package/koa-mount
* https://www.npmjs.com/package/koa-ratelimit
* https://www.npmjs.com/package/koa-rewrite
* https://www.npmjs.com/package/koa-session
* https://www.npmjs.com/package/koa-static
* https://www.npmjs.com/package/@koa/cors

## But seriously, what is this, and how does it work?

You want your server to do a few things for you:

* If the user (tracks $max IPs for $age) has made Too Many Requests ($rpm) throw 429.
* If the middleware chain has been executing for too long ($next) 408 Request Timeout.
* It would be nice if my REST API actually produced error bodies for JSON consumers.
* It would be nice if my logging actually worked how I want it to. (see Bunyan)
* It would be nice if my logs included what I want, and none of what I don't.
* It would be nice if the defaults made sense, and nothing got in my way.
* It would be nice if I could later customize everything how I like it.

Sweet. So, how's it work? See `defaults.js`. Swish.


CHANGELOG.md
# omnibus (koa middleware)

Every published release will now include a changelog entry.

A `git` tag will be dropped at v1.0.0 when that's released.

## v1.0.0

Target release: late 2018

* Add support for an Array of `koa-router` passed as `targetRouters`
* Add support for using `ioredis` to back rate-limits in far-cache
* Add tools for maintainer(s) to best manage releases post v1.0.0

## Pre-v1.0.0

WARNING: These versions may break backward compatibility.

### v0.9.4: added CHANGELOG.md

* Bumped dependencies to latest to fix e.g. `lodash` vulnerability (low)


tests.js
/* eslint-env mocha, node */
const assert = require('assert')
const HTTP = require('http')

const Boom = require('boom') // errors
const supertest = require('supertest')

const omnibus = require('.')

describe('omnibus', () => {

	const test = {}

	const validRequestUUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/
	const validResponseTime = /^[0-9]\.[0-9]{6} millisecond\(s\)$/
	const validDateString = /^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}\.[0-9]{3}Z$/

	describe('usage', () => {

		before(() => {
			const application = omnibus.createApplication(/* options */)
			// same as: `(new (require('koa'))()).use(omnibus(options))`
			application.use((context) => { context.status = 204 })
			test.server = HTTP.createServer(application.callback())
		})

		it('works', () => {
			assert(typeof omnibus === 'function', 'omnibus is not a Function')
			return supertest(test.server)
				.get('/')
				.expect('x-request-id', validRequestUUID)
				.expect('x-response-time', validResponseTime)
				.expect(204, '') // body of No Content
		})

		after(() => {
			test.server.close()
			delete test.server
		})

	})

	describe('errors', () => {

		// 500 Internal Server Error, then 429 Too Many Requests
		const options = { limits: { rpm: 1 }, namespace: false }

		before(() => {
			const application = omnibus.createApplication(options)
			application.use(async () => { throw new Error('test') })
			test.server = HTTP.createServer(application.callback())
		})

		it('returns the proper (clean) JSON payload (500s)', () => {
			const body = Boom.internal().output.payload
			return supertest(test.server)
				.get('/')
				.expect(500, body)
		})

		it('features a simple in-memory rate limiter (429s)', () => {
			const header = `${options.limits.rpm} request(s) per minute`
			const message = `Exceeded rate limit [${header}]` // Error:
			const body = Boom.tooManyRequests(message).output.payload
			return supertest(test.server)
				.get('/')
				.expect('x-ratelimit-limit', header)
				.expect('x-ratelimit-remaining', '0 request(s)')
				.expect('x-ratelimit-reset', validDateString)
				.expect('retry-after', '60')
				.expect(429, body)
		})

		after(() => {
			test.server.close()
			delete test.server
		})

	})

})


demo.js
/* eslint-env node */
const HTTP = require('http')
const URL = require('url')

const omnibus = require('.')

const { getLogger } = require('./log.js')

/* istanbul ignore next */
if (!module.parent) {
	const hostname = process.env.BIND || 'localhost'
	const port = process.env.PORT || 8080 // default
	const application = omnibus.createApplication({
		// allow each client (track 600) 60 req/min (<6s RTT)
		limits: { age: 60000, max: 600, next: 6000, rpm: 60 },
		//limitRate: () => async function noop () {}, // no RL
	})
	const log = getLogger({ component: 'demo' })
	application.use(async function throw500 () {
		throw new Error('demo') // not exposed
	})
	HTTP.createServer(application.callback())
		.listen(port, hostname, () => {
			const serverURL = URL.format({ hostname, port, protocol: 'http:' })
			log.info(`curl -s ${serverURL} -v | jq`) // tests single round-trip
		})
}


index.js
/* eslint-env node */
const KoaApplication = require('koa')

const defaults = require('./defaults.js')

const createOptions = (...args) => {
	//const options = _.defaultsDeep({}, defaults, ...args) // < this doesn't work
	const options = Object.assign({}, defaults, ...args) // however, this works fine:
	options.headers = Object.freeze(Object.assign({}, defaults.headers, options.headers))
	options.limits = Object.freeze(Object.assign({}, defaults.limits, options.limits))
	options.namespace = options.namespace || 'state' // not default namespace
	return Object.freeze(options) // XXX: should Options be a class?
}

const createMiddleware = (options) => {
	const { tracking, timing } = options.headers
	const limitRate = options.limitRate(options)
	const limitTime = options.limitTime(options)
	return async function omnibus (context, next) {
		const headers = options.targetHeaders(options, context, tracking) // Object
		const log = options.targetLogger(options, context, { tracking: headers })
		const source = { log, timing: {}, tracking: headers } // wrapped into:
		const target = options.targetObject(options, context, source)
		const hrtime = target.timing.start = process.hrtime()
		try {
			await limitRate(context, next)
			await limitTime(context, next)
			target.error = null // no throw
		} catch (error) {
			target.error = options.targetError(options, context, error)
		} finally {
			const [s, ns] = process.hrtime(hrtime) // since start
			const ms = Number((s * 1e3) + (ns / 1e6)).toFixed(6)
			headers[timing] = `${ms} millisecond(s)` // tracking
			const err = options.redactedError(options, context)
			const req = options.redactedRequest(options, context)
			const res = options.redactedResponse(options, context)
			if (!err) log.trace({ req, res }, 'request') // softer
			else log.debug({ err, req, res }, 'request') // louder
		}
	}
}

const omnibus = (...args) => {
	const options = createOptions(...args)
	return createMiddleware(options)
}

const createApplication = (...args) => {
	const application = new KoaApplication()
	application.use(omnibus(...args))
	return application
}

module.exports = Object.assign(omnibus, {
	createApplication,
	default: omnibus,
})


defaults.js
/* eslint-env node */
const Boom = require('boom')
const Bunyan = require('bunyan')
const LRU = require('lru-cache')
const UUID = require('uuid')
const _ = require('lodash')

const PACKAGE_NAME = 'omnibus' // make sure to keep DRY
const debug = require('debug')(`koa-${PACKAGE_NAME}`)

const rateLimitByIP = ({ age, max, namespace, rpm }) => {
	const cache = new LRU({ max, maxAge: age })
	const touch = ({ ip: key }) => {
		debug('tracking bucket: %s', key)
		if (!cache.has(key)) {
			const reset = Date.now() + age
			cache.set(key, { count: 0, reset })
		}
		const value = cache.get(key)
		value.count += 1
		return value
	}
	const trackingHeaders = (context, headers) => {
		const target = _.get(context, namespace, {})
		const tracking = _.get(target, 'tracking', {})
		return Object.assign(tracking, headers)
	}
	const middleware = async function throw429 (context) {
		const { count, reset } = touch(context)
		const remaining = Math.max(0, rpm - count)
		const headers = trackingHeaders(context, {
			'X-RateLimit-Limit': `${rpm} request(s) per minute`,
			'X-RateLimit-Remaining': `${remaining} request(s)`,
			'X-RateLimit-Reset': new Date(reset).toISOString(),
		})
		if (rpm < count) {
			headers['Retry-After'] = Math.ceil((reset - Date.now()) / 1000).toFixed(0)
			const message = `Exceeded rate limit [${headers['X-RateLimit-Limit']}]`
			throw Boom.tooManyRequests(message) // data w/ headers is redundant
		}
	}
	return Object.assign(middleware, { cache, touch })
}

const timeBoundedAsyncFunction = (ms, fn) => new Promise((fulfill, reject) => {
	// conforms to https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408:
	const message = `Exceeded time limit [${ms} millisecond(s)]` // see renderBoom
	const error = Boom.clientTimeout(message, { headers: { Connection: 'close' } })
	debug('timeout set: %dms', ms) // do this after?
	const timeout = setTimeout(reject, ms, error)
	fn()
		.then((result) => {
			debug('timeout end: %dms', ms)
			clearTimeout(timeout)
			fulfill(result)
		})
		.catch((reason) => {
			debug('timeout hit: %dms', ms)
			clearTimeout(timeout)
			reject(reason)
		})
})

const getLogger = _.once(() => {
	return Bunyan.createLogger({
		level: process.env.LOG_LEVEL || 'debug',
		name: process.env.LOG_NAME || PACKAGE_NAME,
		serializers: Bunyan.stdSerializers, // Object
		src: process.env.NODE_ENV === 'development',
	})
})

const createBoom = (context, error) => {
	if (error) return Boom.boomify(error, { context })
	return new Boom(`${context.method} ${context.url}`, {
		statusCode: context.status,
	})
}

const renderBoom = (context, namespace) => {
	const object = _.get(context, namespace, {})
	const setHeaders = (parent, keyspace) => {
		const child = _.get(parent, keyspace, {})
		for (const [key, value] of Object.entries(child)) {
			debug('rendered header: %s=%s', key, value)
			context.set(key, value)
		}
	}
	const error = _.get(object, 'error')
	if (error && error.isBoom) {
		const { data, output } = error
		context.status = output.statusCode
		setHeaders(output, 'headers')
		context.body = output.payload
		setHeaders(data, 'headers')
		const summary = output.payload.error
		debug('rendered Boom: %s', summary)
	}
	setHeaders(object, 'tracking')
	return error
}

const trackingObject = (context, key) => {
	const value = context.get(key) || UUID.v1()
	debug('tracking header: %s=%s', key, value)
	return { [key]: value } // more added later
}

const namespacedObject = (context, namespace, ...args) => {
	const error = createBoom(context) // default: 404 Not Found
	context[namespace] = Object(context[namespace]) // decorated:
	return Object.assign(context[namespace], { error }, ...args)
}

const defaults = {
	headers: Object.freeze({ timing: 'X-Response-Time', tracking: 'X-Request-ID' }),
	limits: Object.freeze({ age: 60000, max: 1000 * 1000, next: 60000, rpm: 1000 }),
	namespace: PACKAGE_NAME, // if false-y, will set properties on context.state directly
	limitRate: ({ namespace, limits }) => rateLimitByIP(Object.assign({ namespace }, limits)),
	limitTime: ({ limits }) => (context, next) => timeBoundedAsyncFunction(limits.next, next),
	redactedError: ({ namespace }, context) => renderBoom(context, namespace), // wraps 500s, etc.
	redactedRequest: (options, context) => _.omit(context.request, ['header']), // no Authorization
	redactedResponse: (options, context) => _.omit(context.response, ['header']), // no Set-Cookie
	targetError: (options, context, error) => createBoom(context, error), // [namespace].error Object
	targetHeaders: (options, context, string) => trackingObject(context, string), // tracking headers
	targetLogger: (options, context, object) => getLogger().child(object), // log w/ tracking headers
	targetObject: ({ namespace }, context, source) => namespacedObject(context, namespace, source),
}

module.exports = defaults
